kola/tests/misc/files.go

```diff
@@ -0,0 +1,152 @@
+	"strings"
+
+	"github.com/coreos/mantle/kola/register"
+	"github.com/coreos/mantle/platform"
+)
+
+func init() {
+	register.Register(&register.Test{
+		Run:         SUIDFiles,
+		ClusterSize: 1,
+		Name:        "coreos.filesystem.suid",
+		UserData:    `#cloud-config`,
+	})
+	register.Register(&register.Test{
+		Run:         SGIDFiles,
+		ClusterSize: 1,
+		Name:        "coreos.filesystem.sgid",
+		UserData:    `#cloud-config`,
+	})
+	register.Register(&register.Test{
+		Run:         WritableFiles,
+		ClusterSize: 1,
+		Name:        "coreos.filesystem.writablefiles",
+		UserData:    `#cloud-config`,
+	})
+	register.Register(&register.Test{
+		Run:         WritableDirs,
+		ClusterSize: 1,
+		Name:        "coreos.filesystem.writabledirs",
+		UserData:    `#cloud-config`,
+	})
+}
+
+func sugidFiles(m platform.Machine, validfiles []string, mode string) error {
+	badfiles := make([]string, 0, 0)
+
+	command := fmt.Sprintf("sudo find / -path /sys -prune -o -path /proc -prune -o -type f -perm -%v -print", mode)
+
+	output, err := m.SSH(command)
+	if err != nil {
+		return fmt.Errorf("Failed to run find: output %v, status: %v", output, err)
+	}
+
+	if string(output) == "" {
+		return nil
+	}
+
+	files := strings.Split(string(output), "\n")
+	for _, file := range files {
+		var valid bool
+
+		for _, validfile := range validfiles {
+			if file == validfile {
+				valid = true
+			}
+		}
+		if valid != true {
+			badfiles = append(badfiles, file)
+		}
+	}
+
+	if len(badfiles) != 0 {
+		return fmt.Errorf("Unknown SUID or SGID files found: %v", badfiles)
+	}
+
+	return nil
+}
+
+func SUIDFiles(c platform.TestCluster) error {
+	m := c.Machines()[0]
+
+	validfiles := []string{
+		"/usr/sbin/mount.nfs",
+		"/usr/sbin/unix_chkpwd",
+		"/usr/lib64/polkit-1/polkit-agent-helper-1",
+		"/usr/bin/gpasswd",
+		"/usr/bin/umount",
+		"/usr/bin/ksu",
+		"/usr/bin/newgrp",
+		"/usr/bin/man",
+		"/usr/bin/sudo",
+		"/usr/bin/su",
+		"/usr/bin/passwd",
+		"/usr/bin/expiry",
+		"/usr/bin/chage",
+		"/usr/bin/mandb",
+		"/usr/bin/chsh",
+		"/usr/bin/pkexec",
+		"/usr/bin/mount",
+		"/usr/bin/chfn",
+		"/usr/libexec/dbus-daemon-launch-helper",
+	}
+
+	return sugidFiles(m, validfiles, "4000")
+}
+
+func SGIDFiles(c platform.TestCluster) error {
+	m := c.Machines()[0]
+
+	validfiles := []string{
+	}
+
+	return sugidFiles(m, validfiles, "2000")
+}
+
+func WritableFiles(c platform.TestCluster) error {
+	m := c.Machines()[0]
+
+	output, err := m.SSH("sudo find / -regextype posix-extended -regex \"/(sys|proc)\" -prune -o -type f -perm -0002 -print")
+	if err != nil {
+		return fmt.Errorf("Failed to run find: output %v, status: %v", output, err)
+	}
+
+	if string(output) != "" {
+		return fmt.Errorf("Unknown writable files found: %v", output)
+	}
+
+	return nil
+}
+
+func WritableDirs(c platform.TestCluster) error {
+	m := c.Machines()[0]
+
+	output, err := m.SSH("sudo find / -regextype posix-extended -regex \"/(sys|proc)\" -prune -o -type d -perm -0002 -a ! -perm -1000 -print")
```

Comment:
How many sticky-bit dirs are there? can they be whitelisted instead of universally ignoring anything with the sticky bit?